### ISSUE TITLE
Operators == and != should report incomparable types error

### ISSUE THREAD

**Issue by sourceforge-issue-exporter** [beanshell/beanshell#262]
Heading: The relational operators `==` and `!=`  don't work as per the java syntax rules

Bug Recreation: The following code should report back an error because of the invalid comparison between double and String object, in the 3rd line `scr05[0] == "test"`.

Code:
```java
double[] scr05 = new double[1];
scr05[0] = 1.0;
scr05[0] = ( scr05[0] == "test" ) ? -2 : 1;
```

Possible fix:

Modified `bsh.BSHBinaryExpression` class to handle `==` and `!=` operators as per the java syntax listed below.  This fix worked,  but not very sure whether it could introduce some other bugs in the code. Could you please provide a patch for this, if not at least give us code review feedback?

Java Syntax:
Case 1: If neither LHS operand nor RHS operand is `null`, it directly goes to else condition and checks whether LHS and RHS operands are of same type, if  not reports back an error.

Case 2: If either LHS operand or RHS operand is `null`,  it checks whether LHS or RHS operand is of type primitive, if yes report back an error

Case 3: If both LHS operand and RHS operand are `null`,  no error is reported back (Ex: `if ( null == null) ....`)

Suggested code change in BeanShell interpreter:

Class: `bsh.BSHBinaryExpression`
Method:
```java
public Object eval( CallStack callstack, Interpreter interpreter ) throws EvalError
{
    case EQ:
        if ( ( lhs == Primitive.NULL ) || ( rhs == Primitive.NULL )   )
        {
            if ( ( lhs instance of   Primitive ) && (rhs instanceof   Primitive ) )
                if ( ! ( lhs.getClass().getName().equals(rhs.getClass().getName()) ) )
                {
                    throw new EvalError("Operator: '" + tokenImage[kind] + "' inappropriate for objects", this);
                }
            }
            else
            {
                if ( !( lhs.getClass().getName().equals(rhs.getClass().getName()) ) )
                {
                    throw new EvalError("Operator: '" + tokenImage[kind] + "' inappropriate for objects", this);
                }
        }
        return new Primitive( (lhs == rhs) );
    case NE:
        //same code
        if ( ( lhs == Primitive.NULL ) || ( rhs == Primitive.NULL )   )
        {
            if ( ( lhs instance of   Primitive ) && (rhs instanceof   Primitive ) )
                if ( ! ( lhs.getClass().getName().equals(rhs.getClass().getName()) ) )
                {
                    throw new EvalError("Operator: '" + tokenImage[kind] + "' inappropriate for objects", this);
                }
            }
            else
            {
                if ( !( lhs.getClass().getName().equals(rhs.getClass().getName()) ) )
                {
                    throw new EvalError("Operator: '" + tokenImage[kind] + "' inappropriate for objects", this);
                }
        }
        return new Primitive( (lhs != rhs) );
}
```
thanks,
Lakshmi.

Reported by: *anonymous

Original Ticket: "beanshell/bugs/124":https://sourceforge.net/p/beanshell/bugs/124

**Comment by opeongo** [comment]
This is still a parsing problem.  Here is a simplified example that triggers the error:

```java
double[] d = new double[] {1};
boolean b = (d[0] == "test");
// Error: Parser Error: Unable to parse code syntax. Encountered: [ 0 at line 2, column 15
```

**Comment by jimjag** [comment]
With BeanShell 3.0.0-SNAPSHOT.841 this is now

```
// Error: Parser Error: Unable to parse code syntax. Encountered: [ 0 at line 3, column 15
bsh % // Error: Parser Error: Unable to parse code syntax. Encountered: == at line 1, column 2
bsh % // Error: Parser Error: Unable to parse code syntax. Encountered: ) at line 1, column 8
```

**Comment by nickl-** [comment]
The parser error is a separate issue, this is valid java and should parse.

What the OP is actually after is for beanshell to report an incomparable types error.

Ex:
```java
jshell> new Double(1) == "test"
|  Error:
|  incomparable types: java.lang.Double and java.lang.String
|  new Double(1) == "test"
|  ^---------------------^
```

**Comment by nickl-** [comment]
According

**Comment by nickl-** [comment]
The current implementation will do an object to object comparison and return false. (runtime exception)

**Comment by nickl-** [comment]
I believe the parser error is resolved:

```java
BeanShell 3.0.0-SNAPSHOT.2151
bsh % double[] d = new double[] {1};
--> $0 = {1.0d} :double[]
bsh % boolean b = (d[0] == "test");
--> $1 = false :boolean
```

**Comment by nickl-** [comment]
Can close as duplicate of or to be resolved at #556
